feat: style forgot password form

danielgent opened the pull request:
Check out the URL this ain't login2 :sunglasses: 

![image](https://github.com/sesamyab/auth/assets/8496063/74ae11b8-ef5a-4fd6-8d94-7284cd7ee7cf)

danielgent commented on scripts/upload-css.sh:
```diff
@@ -1,2 +1,5 @@
+yarn tailwind
 npx wrangler r2 object put auth-templates/templates/static/stylesheets/tailwind.css -f src/styles/output.css
+# note inconsistent naming! is served from /css/default.css, is uploaded to tailwind.css, is built as output.css
+# Call it one name everywhere!
```
I'll fix this in another PR now :sunglasses:

danielgent replied:
https://github.com/sesamyab/auth/pull/585

danielgent commented on src/routes/tsx/routes.tsx:
```diff
@@ -6,18 +6,73 @@ import validatePassword from "../../utils/validatePassword";
+  const tenantNameInVendorStyles = tenant.name.toLowerCase();
+  const vendorSettings = await env.fetchVendorSettings(
+    tenantNameInVendorStyles,
+  );
```
this is the main change here :smile: 

@markusahlstrand won't lie this but he said to get everything we need from the tenants table here in auth2 (on planetscale) but it doesn't have most of the fields we need :smile:

markusahlstrand replied:
What are we missing? Guess it's more or less the background image?

danielgent replied:
the vendor settings response has
```
"primaryColor":"#4F3985","buttonTextColor":"#ffffff","primaryHoverColor":"#5F44A0"}
```

Where as the tenants table only has primary & secondary colours

danielgent commented on src/routes/tsx/routes.tsx:
```diff
@@ -6,18 +6,73 @@ import validatePassword from "../../utils/validatePassword";
+
+  const vendorSettings = await env.fetchVendorSettings(
+    tenantNameInVendorStyles,
+  );
```
same change as above. but I moved a lot of these lines up from below as they're needed now to get the vendor settings

danielgent commented on src/routes/tsx/routes.tsx:
```diff
@@ -55,32 +110,31 @@ export async function postResetPassword(
+    return ctx.html(
+      <ResetPasswordPage
+        error="Passwords do not match"
+        vendorSettings={vendorSettings}
```
:eyes: maybe prop drilling isn't the answer *BUT* it will work. we don't have react context here though (nor do we want it!)

markusahlstrand replied:
Could the ResetPasswordPage do it's own data fetching? Not sure if it would be better or worse :)

danielgent replied:
It would still be prop drilling to get the vendor settings into the layout

And we don't want to be fetching the vendor settings *multiple* times :shrug:

danielgent commented on src/server.ts:
```diff
@@ -29,6 +30,17 @@ const server = {
+          const vendorSettings =
+            (await vendorSettingsRes.json()) as VendorSettings;
+
+          return vendorSettings;
```
inter-service dependency hell! :smile:

markusahlstrand replied:
Would be nicer if it was stored in auth2 :) But it's ok to do it in a separate PR..
https://auth0.com/docs/api/management/v2/branding/patch-branding-theme

danielgent replied:
I don't see how a PR here could do this... surely we have to make wide-reaching changes to HQ? It's outside my scope I think :sunglasses:

danielgent commented on src/utils/components/AppLogo.tsx:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  // can't return undefined in hono/jsx so just empty div
+  return <div />;
```
interesting slight change in API

*BUT* this could be a types thing *AND* we're on hono 3. Hono 4 brought lots of JSX improvements so I don't want to judge hono from this!

frankjonatanschultz replied:
just a thought, could you do `return <></>` ?

danielgent replied:
Addressed in https://github.com/sesamyab/auth/pull/589

danielgent commented on src/utils/components/AppLogo.tsx:
```diff
@@ -0,0 +1,25 @@
+      <div class="flex h-9 items-center">
+        <img
+          src={vendorSettings.logoUrl}
+          class="max-h-full"
```
More time burned because this is `class` not `className` here :smile:

danielgent replied:
Copy-pasting straight from login2 and then thinking I've got some serious tailwind build problem *BUT NO*

I'm really surprised this works as `class` is a reserved keyword in JS, thus why React had to do this awfulness

danielgent commented on src/utils/components/Layout.tsx:
```diff
@@ -1,13 +1,27 @@
+const Layout: FC<LayoutProps> = ({ title, children, vendorSettings }) => {
+  const inlineStyles = {
+    backgroundImage: `url(${
+      vendorSettings?.loginBackgroundImage || DEFAULT_BG
```
next PR will be more of this madness then until it renders the same as login2 :smile:

markusahlstrand reviewed (approved):
Looks good :)

frankjonatanschultz commented on src/routes/tsx/routes.tsx:
```diff
@@ -6,18 +6,73 @@ import validatePassword from "../../utils/validatePassword";
+    throw new HTTPException(400, { message: "Tenant not found" });
+  }
+
+  const tenantNameInVendorStyles = tenant.name.toLowerCase();
```
I think this should be `session.authParams.client_id`..?

Just thinking if the vendor has a name with a space, like this one:
https://api.sesamy.com/profile/vendors/billgren-wood/style

`name` is `Billgren Wood` but you want this to be `billgren-wood`

danielgent replied:
I've no idea! I'll do as you requested

On login2 we have `vendor_id` for the styling, and `client_id` for the actual auth0/2 OAuth client

*but* we must only have the client id here

so we'll lose the styling somewhere but we're only going to release this for kvartal initially so let's cross that bridge when we come to it :smile:

danielgent replied:
Addressing here https://github.com/sesamyab/auth/pull/589

danielgent commented:
@frankjonatanschultz I've been spending too long programming with Markus! Apologies. I just merged the PR without addressing your comments! :smile: 

They're very legit. I'll address these now on new PRs.  

This makes you a fullstack dev as well :sunglasses:
